syntax = "proto3";


service Raft {
  rpc Vote (RequestVoteRPC) returns (ResponseVoteRPC) {}
  rpc AppendMessage (RequestAppendEntriesRPC) returns (ResponseAppendEntriesRPC) {}
}

message RequestVoteRPC {
  int32 term = 1;
  int32 candidateId = 2;
  int32 lastLogIndex = 3;
  int32 lastLogTerm = 4;
}

message ResponseVoteRPC {
  int32 term = 1;
  bool voteGranted = 2;
}

message RequestAppendEntriesRPC {
  int32 term = 1;
  int32 leaderId = 2;
  int32 prevLogIndex = 3;
  int32 prevLogTerm = 4;
  LogEntry entry = 5;
  int32 leaderCommit = 6;
}

message ResponseAppendEntriesRPC {
  int32 term = 1;
  bool success = 2;
}

message LogEntry {
  int32 term = 1;
  string command = 2;
}
